### PR TITLE
[WIP] mlp6-review-me

### PR DESCRIPTION
**Pull request by cshackerpro**

### CODE REVIEW

File: ecg_data.py [hunk 1]
```diff
@@ -1,12 +1,100 @@
-class hrm:
+import numpy
+import pandas
+
+
+class ecg_data:
```
Comment (mlp6) [inline ecg_data.py]: PEP8 class naming conventions

File: ecg_data.py [hunk 2]
```diff
@@ -1,12 +1,100 @@
...
     'This is a class'
-    def __init__(self, filename, avgT1,avgT2,bradyT,bradyThresh,tachyT,tachyThresh):
-        self.avgT1 = 0.0
-        self.avgT2 = 10.0
 
-    #def instantHr(self):
+    def __init__(
```
Comment (mlp6) [inline ecg_data.py]: Need docstring, and specify units for some input parameters.

File: ecg_data.py [hunk 3]
```diff
@@ -1,12 +1,100 @@
...
+            self,
+            filename,
+            begin_time=0,
+            end_time=10,
+            bradyT=5,
+            bradyThresh=60,
+            tachyT=5,
+            tachyThresh=100):
+
```
Comment (mlp6) [inline ecg_data.py]: Do you not want to map any of the input variables to object attributes?  (I think you do!)  ;)

File: ecg_data.py [hunk 4]
```diff
@@ -1,12 +1,100 @@
...
+        # put peak detection here
+        data = pandas.read_csv(
```
Comment (mlp6) [inline ecg_data.py]: This should be broken out into its own method.

File: ecg_data.py [hunk 5]
```diff
@@ -1,12 +1,100 @@
...
+            filename, converters={
+                "times": float, "voltages": float})
+        voltages = data.voltages.values
+        finalTimes = [0]
```
Comment (mlp6) [inline ecg_data.py]: ??  not clear what `finalTime` represents; would try to make this variable name more self-descriptive.

Comment (mlp6) [reply ecg_data.py]: Why is this 0?  Would `None` be more appropriate?

File: ecg_data.py [hunk 6]
```diff
@@ -1,12 +1,100 @@
...
+        avgvoltage = numpy.average(voltages)
+
+        threshvoltage = avgvoltage * 2
```
Comment (mlp6) [inline ecg_data.py]: Where does `2` come from?

File: ecg_data.py [hunk 7]
```diff
@@ -1,12 +1,100 @@
...
+
+        peaks = numpy.where(voltages >= threshvoltage)
+        peaks1 = peaks[0]
```
Comment (mlp6) [inline ecg_data.py]: The use of these placeholder variables (`peaks1` and `peaks2`) helps suggest that this section of code should also be broken out into its own method.

File: ecg_data.py [hunk 8]
```diff
@@ -1,12 +1,100 @@
...
+        peaks2 = peaks1.tolist()
+        boxcar = []
+        for i in range(1, len(peaks2) - 1):
```
Comment (mlp6) [inline ecg_data.py]: The logic of what is going on here is hard to follow!  Would try to create some new methods that are descriptive of what they are doing.

File: ecg_data.py [hunk 9]
```diff
@@ -1,12 +1,100 @@
...
+                    (voltages[peaks2[i]] >= voltages[peaks2[i + 1]]):
+                recentval = finalTimes[len(finalTimes) - 1]
+                finalTimes.append(peaks2[i])
+                if (peaks2[i] - recentval <= 50):
+                    finalTimes.pop()
+        finalTimes.pop(0)
+
+        self.times = finalTimes
+        self.inst = self.instantHr(begin_time)
+        self.avg = self.averageHR(begin_time, end_time)
```
Comment (mlp6) [inline ecg_data.py]: would use some more descriptive attribute names here (don't need to be too brief)

File: ecg_data.py [hunk 10]
```diff
@@ -1,12 +1,100 @@
...
+        self.ano = self.anomalyHr(bradyT, bradyThresh, tachyT, tachyThresh)
```
Comment (mlp6) [inline ecg_data.py]: if all of these inputs were attributes, they would be available  "for free" as part of `self`

File: ecg_data.py [hunk 11]
```diff
@@ -1,12 +1,100 @@
...
+
+    def instantHr(self, target_time = 0):
```
Comment (mlp6) [inline ecg_data.py]: why is `target_time` not an attribute set in `__init__()`?

File: ecg_data.py [hunk 12]
```diff
@@ -1,12 +1,100 @@
...
+        for x in range(0, len(self.time)):
+            if self.time[x] > target_time:
+                instant_dt = self.time[x + 1] - self.time[x]
+                break
+
+        return (60 / instant_dt) * 100
```
Comment (mlp6) [inline ecg_data.py]: why not assign to an object attribute?  this is not a static method; not sure what is going on here.

File: ecg_data.py [hunk 13]
```diff
@@ -1,12 +1,100 @@
...
+
+    def averageHr(self, begin_time = 0, end_time = 10):
```
Comment (mlp6) [inline ecg_data.py]: again, why are `begin_time` and `end_time` not attributes?  (maybe there is a good reason... not obvious to me)

File: ecg_data.py [hunk 14]
```diff
@@ -1,12 +1,100 @@
...
+        for j in range(1, len(self.time)):
```
Comment (mlp6) [inline ecg_data.py]: really hard to follow the logic of what is going on here

File: ecg_data.py [hunk 15]
```diff
@@ -1,12 +1,100 @@
...
+                    (self.time[j] / 1000 > begin_time):
+                begin = j
+            if (self.time[j - 1] / 1000 == end_time):
+                end = j - 1
+            elif (self.time[j - 1] / 1000 < end_time) and \
+                    (self.time[j] / 1000 > end_time):
+                end = 1
+        time_count = 0
+
+        for k in range(begin + 1, end + 1):
```
Comment (mlp6) [inline ecg_data.py]: also not sure what is going on here...

File: ecg_data.py [hunk 16]
```diff
@@ -1,12 +1,100 @@
...
+            time_count = time_count + (self.time[k] - self.time[k - 1]) / 1000
+
+        div = end - begin
 
-    #def averageHr(self):
+        time_avg = time_count / div
+        return 60 / time_avg
```
Comment (mlp6) [inline ecg_data.py]: set attribute?

File: ecg_data.py [hunk 17]
```diff
@@ -1,12 +1,100 @@
...
+            if (60000 / (self.time[l] - self.time[l - 1])) < \
+                    (tachyThresh and dying_fast == 0):
+                dying_fast = self.time[l - 1]
+            elif (dying_fast != 0) and \
+                    (60000 / (self.time[l] - self.time[l - 1]) < tachyThresh):
+                if self.time[l] - dying_fast > tachyT:
+                    tachyTimes.append(dying_fast / 1000)
+                dying_fast = 0
 
+        return bradyTimes, tachyTimes
```
Comment (mlp6) [inline ecg_data.py]: attributes?

File: ecg_data.py [hunk 18]
```diff
@@ -1,12 +1,100 @@
...
+        return 60 / time_avg
 
-    #def anomalyHr(self):
+    def anomalyHr(self, bradyT = 5, bradyThresh = 60, tachyT = 5, tachyThresh = 100):
+        dying_slow = 0
+        dying_fast = 0
+        bradyTimes = []
+        tachyTimes = []
+        counter = 0
+        for l in range(1, len(self.time)):
```
Comment (mlp6) [inline ecg_data.py]: oh my... definitely would try to refactor this for readability

File: ecg_data.py (hunk 11 above)
Comment (cshackerpro) [reply ecg_data.py]: We wanted the people to be able to update the target time, so wont it need to be an input? Instead of having them update the attribute and run the method again? Or is it better practice for them to do that?

Comment (mlp6) [reply ecg_data.py]: You could use a "setter" method... which would involve updating an attribute value and then automatically triggering all of the dependent methods to be re-run and updated.  Check out: https://www.python-course.eu/python3_properties.php, which highlights the use of the `@property` decorator.

File: hrm.py [hunk 19]
```diff
@@ -1,104 +1,122 @@
 # HRM
 
-import pytest
+
 import numpy
 import pandas
 import scipy.signal
 import matplotlib.pyplot as plt
 
+def fileChecker(ecg_data, names):
```
Comment (suyashkumar) [inline hrm.py]: nit: canonically function names should be snake_case in python. From PEP8:
>Function names should be lowercase, with words separated by underscores as necessary to improve readability.

File: hrm.py [hunk 20]
```diff
@@ -1,104 +1,122 @@
...
+    try:
+        df = pandas.read_csv(ecg_data, header=None, names=names, converters={"times":float,"voltages":float})
+        errBool = False
```
Comment (suyashkumar) [inline hrm.py]: nit: generally snake_case for vars too :)

File: hrm.py [hunk 21]
```diff
@@ -1,104 +1,122 @@
...
         :param ecg_data: the initial csv file given
         :rtype: list of times where a peak occured
     """
+    peakTimes=[0]
 
-    # header is time, voltage
-    data = pandas.read_csv(ecg_data, converters={"times": float, "voltages": float})
-    # avgVoltage = numpy.mean(data.voltages.values)
-    # minVoltage = numpy.min(data.voltages.values)
+    names = ["times","voltages"]
```
Comment (suyashkumar) [inline hrm.py]: nit: add space after the comma

File: hrm.py (hunk 19 above)
Comment (suyashkumar) [reply hrm.py]: also, add docstring :)

File: hrm.py [hunk 22]
```diff
@@ -1,104 +1,122 @@
...
+    dataerror = False
```
Comment (suyashkumar) [inline hrm.py]: snake_case -> `data_error`

File: hrm.py [hunk 23]
```diff
@@ -1,104 +1,122 @@
...
+    dataerror = fileChecker(ecg_data, names)
+    if (dataerror == True):
+        print("Non-Numeric Value Entered")
+        return peakTimes
+
+    data = pandas.read_csv(ecg_data, header=None, names=names, converters={"times":float, "voltages":float})
```
Comment (suyashkumar) [inline hrm.py]: nit in key value pairs in dictionaries, generally there's a space after the `:`

File: hrm.py [hunk 24]
```diff
@@ -1,104 +1,122 @@
...
+def fileChecker(ecg_data, names):
+    try:
+        df = pandas.read_csv(ecg_data, header=None, names=names, converters={"times":float,"voltages":float})
+        errBool = False
+        return errBool
+    except ValueError:
+        errBool = True
+        return errBool
 
 def peakDetector(ecg_data):
```
Comment (suyashkumar) [inline hrm.py]: from the code reviews I was doing before this pull request: #34

File: hrm.py [hunk 25]
```diff
@@ -1,104 +1,122 @@
...
-        if (voltages[peaks2[i]]>=voltages[peaks2[i-1]]) and (voltages[peaks2[i]]>=voltages[peaks2[i+1]]):
-            recentval = finalTimes[len(finalTimes)-1]
-            finalTimes.append(peaks2[i])
-            if(peaks2[i]-recentval<=50):
-                finalTimes.pop()
-    finalTimes.pop(0)
-    return finalTimes
-
+    peaks2 = peaks1.tolist()
+    first_peak = 0.0;
```
Comment (suyashkumar) [inline hrm.py]: nit: semicolons not needed

File: hrm.py [hunk 26]
```diff
@@ -110,46 +128,43 @@ def average(time, begin_time, end_time):
...
         raise ValueError('Begin and End time are too close')
 
     time_avg = time_count / div
 
-    #print (60/time_avg)
-
     return 60 / time_avg
 
 
 def anomaly(time, brady_thresh, brady_time, tachy_thresh, tachy_time):
```
Comment (suyashkumar) [inline hrm.py]: nit: generally name functions using actions instead as a noun or entity (that should be reserved for classess). Something like `find_...` may be apt here

File: hrm.py [hunk 27]
```diff
@@ -171,25 +186,38 @@ def anomaly(time, brady_thresh, brady_time, tachy_thresh, tachy_time):
     tachyTimes = []
     counter = 0
     for i in range(1, len(time)):
-        if 60000 / (time[i] - time[i-1]) < brady_thresh and dying_slow == 0:
+        if (60000 / (time[i] - time[i - 1]) < brady_thresh) and \
```
Comment (suyashkumar) [inline hrm.py]: 60000 should likely be defined as a constant at the top of the file in a descriptive way. an example could be:
```python
SECONDS_IN_MIN = 60
```

File: hrm_class.py [hunk 28]
```diff
@@ -0,0 +1,231 @@
+import numpy
+import pandas
+
+class hrm_data:
```
Comment (suyashkumar) [inline hrm_class.py]: oddly enough in Python (PEP8), classes should be CapCased (so should be `HrmData` or `HRMData`: https://www.python.org/dev/peps/pep-0008/#class-names)

File: hrm_class.py [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+            tachyTimes
+
+    """
+
+    def __init__(
+            self,
+            filename,
+            begin_time=0,
+            end_time=10,
+            bradyT=5,
```
Comment (suyashkumar) [inline hrm_class.py]: as with elsewhere in the code, a reminder to use snake_case and generally be consistent with naming :)

File: hrm_main.py [hunk 30]
```diff
@@ -0,0 +1,16 @@
...
+        user_specified_time1=0,
+        user_specified_time2=30,
+        brady_threshold=50,
+        tachy_threshold=100,
+        brady_time=5,
+        tachy_time=5,
+        inst=False,
+        avg=False,
+        ano=False):
+    hrm_object = hrm_class.hrm_data(filename)
```
Comment (suyashkumar) [inline hrm_main.py]: generally, simply instantiating a class should not be performing the whole analysis. (This is assuming that, in this case, your goal was for the `hrm_data` class to be a container for the raw ECG data. If not, then perhaps consider renaming the class--we can dive deeper into this if you want). Consider instantiating a new HRMData object and then you as the user of your class should use the class API to perform some explicit actions on that data container.
```python
hrm = HRMData(filename)
hrm.calculate_heart_rates()
hrm.write_summary_to_file("test.txt")
...
etc
```
You'll notice this general pattern of using a class throughout all the modules you use in python :)

File: hrm.py (hunk 27 above)
Comment (cshackerpro) [reply hrm.py]: Should we still fix hrm.py if we're not using it in class_based approach?

Comment (cshackerpro) [reply hrm.py]: Also for the class-based, would this be a global variable outside of the class?

File: ecg_data.py (hunk 11 above)
Comment (cshackerpro) [reply ecg_data.py]: Is this new update what you were talking about?

File: ecg_data.py (hunk 14 above)
Comment (cshackerpro) [reply ecg_data.py]: Do you just want me to comment the code? Or do you want me to change the average function?

File: ecg_data.py (hunk 15 above)
Comment (cshackerpro) [reply ecg_data.py]: Same with the comment form above

File: ecg_data.py (hunk 18 above)
Comment (cshackerpro) [reply ecg_data.py]: @syiyuli

File: hrm.py (hunk 27 above)
Comment (suyashkumar) [reply hrm.py]: @cshackerpro focus on getting the class based approach fixed first and then circle back to `hrm.py` to clean things up if you have time (it shouldn't take too long). 

Yep, this would be a global variable outside the class. Among many reasons for doing this, this ensures this constant isn't reallocated in memory for every new instantiation of your class. It also ensures all class instances are using the same constant (less risk of one class changing it's instance variable constant and having others use them, although ideally this would be a static constant but python isn't really into that :( )
